pkg/controllermanager/controller/certificatesigningrequest/reconciler.go

```diff
@@ -0,0 +1,150 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+This file was copied and modified from the kubernetes/kubernetes project
+https://github.com/kubernetes/kubernetes/blob/release-1.15/pkg/controller/certificates/approver/sarapprove.go
+
+Modifications Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
+*/
+
+package certificatesigningrequest
+
+import (
+	"context"
+	"fmt"
+
+	authorizationv1 "k8s.io/api/authorization/v1"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	certificatesclientv1 "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/controllermanager/apis/config"
+	"github.com/gardener/gardener/pkg/utils"
+	gutil "github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+type certificatesClientSet interface {
+	CertificatesV1() certificatesclientv1.CertificatesV1Interface
+}
+
+// Reconciler reconciles CertificateSigningRequest.
+type Reconciler struct {
+	Client             client.Client
+	CertificatesClient certificatesClientSet
+	Config             config.CertificateSigningRequestControllerConfiguration
+}
+
+// Reconcile performs the main reconciliation logic.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var (
+		log = logf.FromContext(ctx)
+
+		cert               []byte
+		isInFinalState     bool
+		finalState         string
+		req                []byte
+		usages             []certificatesv1.KeyUsage
+		extra              = make(map[string]authorizationv1.ExtraValue)
+		username           string
+		uid                string
+		groups             []string
+		updateConditionsFn func() error
+	)
+
+	csr := &certificatesv1.CertificateSigningRequest{}
+	if err := r.Client.Get(ctx, request.NamespacedName, csr); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	for _, c := range csr.Status.Conditions {
+		if c.Type == certificatesv1.CertificateApproved || c.Type == certificatesv1.CertificateDenied {
+			isInFinalState = true
+			finalState = string(c.Type)
+		}
+	}
+	for k, v := range csr.Spec.Extra {
+		extra[k] = authorizationv1.ExtraValue(v)
+	}
+	cert = csr.Status.Certificate
+	req = csr.Spec.Request
+	usages = csr.Spec.Usages
+	username = csr.Spec.Username
+	uid = csr.Spec.UID
+	groups = csr.Spec.Groups
```

Comment:
We don't need these variables anymore not that we only work with `v1`. Let's drop them and access `csr.{Spec,Status}.*` directly.